packages/yarnpkg-core/sources/Controllers.ts

```diff
@@ -0,0 +1,105 @@
+    if (resolver) {
+      const next = this.index.get(resolver); // if next is undefined the `prepareNext` will use the protocol MultiResolver
+      return (resolver as any)[func].bind(resolver, ...args, this.prepareNext(func, next, ...args), this.protocols);
+    } else {
+      return (this.protocols as any)[func].bind(this.protocols, ...args);
+    }
+  }
+
+  async supportsDescriptor(descriptor: Descriptor, opts: MinimalResolveOptions): Promise<boolean> {
+    return await this.prepareNext('supportsDescriptor', this.modules[0], descriptor, opts)();
+  }
+  async supportsLocator(locator: Locator, opts: MinimalResolveOptions): Promise<boolean> {
+    return await this.prepareNext('supportsLocator', this.modules[0], locator, opts)();
+  }
+  async shouldPersistResolution(locator: Locator, opts: MinimalResolveOptions): Promise<boolean> {
+    return await this.prepareNext('shouldPersistResolution', this.modules[0], locator, opts)();
+  }
+  async bindDescriptor(descriptor: Descriptor, fromLocator: Locator, opts: MinimalResolveOptions): Promise<Descriptor> {
+    return await this.prepareNext('bindDescriptor', this.modules[0], descriptor, fromLocator, opts)();
+  }
+  async getResolutionDependencies(descriptor: Descriptor, opts: MinimalResolveOptions): Promise<Descriptor[]> {
+    return await this.prepareNext('getResolutionDependencies', this.modules[0], descriptor, opts)();
+  }
+  async getCandidates(descriptor: Descriptor, dependencies: Map<DescriptorHash, Package>, opts: ResolveOptions): Promise<Locator[]> {
+    return await this.prepareNext('getCandidates', this.modules[0], descriptor, dependencies, opts)();
+  }
+  async resolve(locator: Locator, opts: ResolveOptions): Promise<Package> {
+    return await this.prepareNext('resolve', this.modules[0], locator, opts)();
+  }
+}
+
+export class FetcherController implements Fetcher {
+  public readonly type = 'module';
+  private readonly index = new Map<Fetcher, Fetcher | undefined>();
+  private readonly protocols = new MultiFetcher([]);
+  private readonly modules = new Array<Fetcher>();
+
+  constructor(fetchers: Array<Fetcher>) {
+    fetchers.forEach( (fetcher) => {
+      if (fetcher && fetcher.type === 'module') {
+        this.modules.push(fetcher);
+      } else if (fetcher) {
+        this.protocols.addFetcher(fetcher);
+      }
+    } );
+    for (let i=0; i < this.modules.length; i++) {
+      this.index.set(this.modules[i], this.modules[i+1]);
+    }
+  }
+
+  prepareNext(func: string, fetcher: Fetcher | undefined, ...args: any): any {
+    if (fetcher) {
+      const next = this.index.get(fetcher);
+      return (fetcher as any)[func].bind(fetcher, ...args, this.prepareNext(func, next, ...args), this.protocols);
+    } else {
+      return (this.protocols as any)[func].bind(this.protocols, ...args);
+    }
+  }
+
+  async supports(locator: Locator, opts: import("./Fetcher").MinimalFetchOptions): Promise<boolean> {
```

Comment:
Why using the `import(...)` syntax instead of importing the types in the header?

Reply:
I am not exactly sure what you are referring to, but the answer is probably that I am not very familiar with typescript